File: 2024/Days/1-5/1/Code/Typescript/part1Solution.ts
class Part1Solution{
	// Reads a file and returns its content in utf-8 fencoding.
	readFile = async (file_name: string) => {
		const fs = await import('fs');

		return new Promise<string>((resolve, reject) => {
			fs.readFile(file_name, 'utf-8', (err, data) => {
				if (err) reject(err)

				resolve(data)
			});
		}).catch((error) => {
			if (error.code === 'ENOENT') {
				console.error('File not found');
			} else {
				console.error(`Error: ${error.message}`);
			}
			process.exit(1)
		})
	};


	getMatrix = (content: string) => {
		let column1: Array<number> = []
		let column2: Array<number> = []

		let rows = content.split('\n')
		rows.forEach(row => {
			const values = row.split(' ')

			// Defaults empty lines as 0
			const num1 = parseInt(values[0]) || 0
			const num2 = parseInt(values[values.length - 1]) || 0

			column1.push(num1)
			column2.push(num2)
		})

		return [column1, column2]
	}

	getDifference = (matrix: Array<Array<number>>) => {
		const differenceMap = matrix[0].map((num, index) => Math.abs(num - matrix[1][index]));
		return differenceMap.reduce((accumulator, number) => accumulator + number)
	}
}


(async _ => {
	const part1Solution = new Part1Solution()
	const content: string = await part1Solution.readFile('input1.txt');

	const matrix = part1Solution.getMatrix(content)

	matrix[0].sort()
	matrix[1].sort()

	const totalDifference = part1Solution.getDifference(matrix)
	console.log(totalDifference);
})()

File: 2024/Days/1-5/1/Code/Dart/part1Solution.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:core';


import 'fileParseHandler.dart';


class Part1Solution{
	Future<int> execute() async{
		var parser = new FileParseHandler();
		var matrix = await parser.parseFile<List<List<int>>>("input1.txt", parseInput1);

		// NULL CHECK SAFETY
		if(matrix == null){
			return 0;
		}

		if(matrix.length != 2){
			return 0;
		}

		var column1 = matrix[0];
		var column2 = matrix[1];

		// Sort columns
		column1.sort();
		column2.sort();
		
		// Get total difference and prints the result.
		int totalDifference = getTotalDifference(column1, column2);
		return totalDifference;
	}

	Future<List<List<int>>> parseInput1(Stream<String> lines) async{
		List<int> column1 = [];
		List<int> column2 = [];
		await for(var line in lines){
			if(line.isEmpty){
				continue;
			}
			var lineSplitted = line.split(' ');
			column1.add(int.parse(lineSplitted.first));
			column2.add(int.parse(lineSplitted.last));
		}

		return [column1, column2];
	}


	int getTotalDifference(List<int> column1, List<int> column2){
		int totalDifference = 0;
		for(int i = 0; i < column1.length; i++){
			totalDifference += (column1[i] - column2[i]).abs();
		}
		return totalDifference;
	}
}
